File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        
        l, r   = 0, len(nums) - 1
        left_v = nums[0]
        
        while l < r:
            m     = (l + r) // 2
            m_val = nums[m]

            if   m_val <  left_v <= target:  r = m - 1
            elif m_val >= left_v >  target:  l = m + 1 
            elif m_val <  target:            l = m + 1
            elif m_val >  target:            r = m - 1
            else:                            return m
                    
        if nums[l] == target: return l
        if nums[r] == target: return r
        return -1

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l=0,h=nums.size()-1;
        while(l<=h){
            int mid=(l+h)>>1;
            if(nums[mid]==target) return mid; // when we found the target 
            if(nums[l]<=nums[mid]){ // if selected part is  sorted 
                if(target>=nums[l]&& target<=nums[mid]) // checking if target lies in l-mid
                    h=mid-1;
                else
                    l=mid+1;
            }
            else{
                if(target>=nums[mid]&& target<=nums[h])// checking if target lies in mid-h
                    l=mid+1;
                else 
                    h=mid-1;
            }
        }
        return -1; // when target not found
    }
};